Mathlib/Data/Matrix/Kronecker.lean

```diff
@@ -282,7 +282,8 @@ theorem kronecker_apply [Mul α] (A : Matrix l m α) (B : Matrix n p α) (i₁ i
 /-- `Matrix.kronecker` as a bilinear map. -/
 def kroneckerBilinear [CommSemiring R] [Semiring α] [Algebra R α] :
     Matrix l m α →ₗ[R] Matrix n p α →ₗ[R] Matrix (l × n) (m × p) α :=
-  kroneckerMapBilinear (Algebra.lmul R α)
+  kroneckerMapBilinear (Algebra.lmul R α).toLinearMap
+  -- ACL : why is it neessary to specify .toLinearMap ?
```

Comment:
Let's try to debug this in the hope it helps with the rest of the issues.

On the `master` branch this coercion goes `AlgHom → NonUnitalAlgHom → DistribMulActionHom → LinearMap`. Here it looks like we get to `CoeTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α)` which fails to find the right instance and so we can't find the coercion. On master the trace is:
```lean
  [] ✅ apply @instCoeOTC_1 to CoeOTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α) ▼
    [tryResolve] ✅ CoeOTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α) ≟ CoeOTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α)
    [] new goal CoeTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α) ▼
      [instances] #[@instCoeTC, @instCoeTC_1, @instCoeTC_2, @DistribMulActionHom.instCoeTCDistribMulActionHomToMonoidToMonoidWithZeroToAddMonoidToDistribMulActionToAddMonoidToDistribMulActionLinearMapIdToNonAssocSemiring]
```
on this branch:
```lean
  [] ✅ apply @instCoeOTC_1 to CoeOTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α) ▼
    [tryResolve] ✅ CoeOTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α) ≟ CoeOTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α)
    [] new goal CoeTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α) ▼
      [instances] #[@instCoeTC, @instCoeTC_1, @instCoeTC_2]
```
So the question is why `DistribMulActionHom.instCoeTCDistribMulActionHomToMonoidToMonoidWithZeroToAddMonoidToDistribMulActionToAddMonoidToDistribMulActionLinearMapIdToNonAssocSemiring` isn't considered as possible instance. It turns out the instance is actually changed slightly, `Mathlib/Algebra/Module/LinearMap.lean:677`:
```diff
- instance : CoeTC (M →+[R] M₂) (M →ₗ[R] M₂) :=
+ instance : Coe (M →ₑ+[σ.toMonoidHom] M₂) (M →ₛₗ[σ] M₂) :=
```
so let's instead look at the part of the trace where `Coe` instances are considered:
```lean
  [] ✅ apply @instCoeTC_1 to CoeTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α) ▼
    [tryResolve] ✅ CoeTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α) ≟ CoeTC (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α)
    [] no instances for Coe (α →+[R] Module.End R α) (α →ₗ[R] α →ₗ[R] α) ▼
      [instances] #[]
```
again, no instances! So it seems that the instance we want is somewhere else in the discrimination tree, and therefore cannot be found. My suspicion is that it doesn't see that `M →ₑ+[MonoidHom.id R] M₂` can unify with `M →ₑ+[(?σ).toMonoidHom] M₂`.

So to fix the coercion, I think the most direct way is to add an additional `Coe` instance specialized to the non-semi `DistribMulActionHom` and `LinearMap`. In addition, it is very surprising we don't have a general instance `[SemiLinearMapClass F σ M N] : CoeHead F (M →\sl [σ] N)`, like we do for basically any other of these morphism classes. I think I'll try to add the latter on a new branch off of master and see what breaks.

Reply:
To see which coercions were used on the master branch: use the `#print kroneckerBilinear` command after this definition, and then click on the coercion arrows to see their definition.

To see the typeclass trace: use the `set_option trace.Meta.synthInstance true` before this definition, and match this up with what you expect to see. Here I found it useful to run the same command in another copy of mathlib where i have checked out the master branch, so I could easily compare the output. I searched for each instance we'd expect the search to apply, and if I couldn't find it I looked higher up in the trace to find out why. The instance search algorithm for coercions is a bit hard to follow because of all the `CoeT`, `CoeTC`, `CoeHTCT` classes that have nondescriptively named instances, so I typically resort to trial and error to find what I really care about.

Reply:
Indeed, merging #10758 into this branch seems to fix the issue. I took the liberty of pushing my merge to your branch: feel free to revert!

Reply:
Thanks!